src/Symfony/Component/Cache/Adapter/RedisTagAwareAdapter.php

```diff
@@ -161,46 +143,69 @@ protected function doDelete(array $ids, array $tagData = []): bool
      */
     protected function doInvalidate(array $tagIds): bool
     {
-        if (!$this->redisServerSupportSPOP()) {
+        if ($this->redis instanceof \Predis\ClientInterface && $this->redis->getConnection() instanceof ClusterInterface) {
+            CacheItem::log($this->logger, 'Cluster on Predis does not support RENAME so there is a risk of race conditions on tag invalidation, use phpredis instead.');
+            $movedTagSetIds = $tagIds;
+        } else {
+            $movedTagSetIds = $this->renameKeys($tagIds);
+        }
+
+        // No Sets found
+        if (empty($movedTagSetIds)) {
             return false;
         }
 
-        // Pop all tag info at once to avoid race conditions
-        $tagIdSets = $this->pipeline(static function () use ($tagIds) {
-            foreach ($tagIds as $tagId) {
-                // Client: Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
-                // Server: Redis 3.2 or higher (https://redis.io/commands/spop)
-                yield 'sPop' => [$tagId, self::POP_MAX_LIMIT];
+        // Now safely take the time to read the keys in each set and collect ids we need to delete
+        $tagIdSets = $this->pipeline(static function () use ($movedTagSetIds) {
+            foreach ($movedTagSetIds as $movedTagId) {
+                yield 'sMembers' => [$movedTagId];
             }
         });
 
-        // Flatten generator result from pipeline, ignore keys (tag ids)
-        $ids = array_unique(array_merge(...iterator_to_array($tagIdSets, false)));
+        // Return combination of the temporary Tag Set id's and the their values (cache id's)
+        $ids = array_merge($movedTagSetIds, ...iterator_to_array($tagIdSets, false));
 
         // Delete cache in chunks to avoid overloading the connection
-        foreach (array_chunk($ids, self::BULK_DELETE_LIMIT) as $chunkIds) {
+        foreach (array_chunk(array_unique($ids), self::BULK_DELETE_LIMIT) as $chunkIds) {
             $this->doDelete($chunkIds);
         }
 
         return true;
     }
 
-    private function redisServerSupportSPOP(): bool
+    /**
+     * Rename several keys in order to be able to operate on them without risk of race conditions.
+     *
+     * Filters out keys that does not exist before returning new keys.
+     *
+     * @see https://redis.io/commands/rename
+     *
+     * @param array $ids The original cache ids
```

Comment:
let's remove this it's mostly useless

Reply:
```suggestion
```

Reply:
👎 That won't work for this one, I'll take that onelocallyy ;)